### PR TITLE
[NuclioRuntime] Always pass serving graph to pod using config-map

### PR DESCRIPTION
**Pull request by alxtkr77**
https://iguazio.atlassian.net/browse/ML-6223

### CODE REVIEW

File: mlrun/utils/helpers.py [hunk 1]
```diff
@@ -1550,12 +1550,6 @@ def get_serving_spec():
     except FileNotFoundError:
         pass
 
-    if data is None:
-        data = os.environ.get("SERVING_SPEC_ENV", "")
```
Comment (gtopper) [inline mlrun/utils/helpers.py]: Can we keep this fallback? I use it quite extensively for running functions locally, and having to create a file does complicate things a bit, especially when its path is hard coded.

File: tests/serving/test_serving.py [hunk 2]
```diff
@@ -36,6 +37,16 @@
 from mlrun.utils import logger
 
 
+def setUpModule():
+    with tempfile.NamedTemporaryFile(mode="w+", delete=False) as temp_file:
+        mlrun.common.constants.MLRUN_SERVING_SPEC_PATH = temp_file.name
```
Comment (gtopper) [inline tests/serving/test_serving.py]: I wonder if the path should be properly overridable via an environment variable. This avoids the hacks in the tests/benchmark and also can be helpful in debugging.

File: tests/serving/test_serving.py [hunk 3]
```diff
@@ -36,6 +37,16 @@
...
+
+
+# Global teardown function
+def tearDownModule():
```
Comment (gtopper) [inline tests/serving/test_serving.py]: Redundant comment that's inconsistent with `setUpModule()`.
```suggestion
def tearDownModule():
```

**Review by gtopper — COMMENTED** [review]
Looks good, but I'm wondering if we can keep the flexibility so we're not married to a hard coded path under `/tmp`.

File: mlrun/utils/helpers.py (hunk 1 above)
Comment (alxtkr77) [reply mlrun/utils/helpers.py]: While I understand the reasoning, I personally oppose having two methods for doing the same thing. Regarding testing, you can always override `mlrun.common.constants.MLRUN_SERVING_SPEC_PATH` with a temporary file. I will add your comment to the case and see what others have to say.

Comment (gtopper) [reply mlrun/utils/helpers.py]: I don't see a problem with keeping the override, but anyway, see my other comment regarding overwriting the global constant. As far as I can tell, it's not done anywhere else in the code base.

**Review by theSaarco — CHANGES_REQUESTED** [review]
@alxtkr77 - looking at this, I actually think there is an issue here with BC. If the user is running a function that runs with an older mlrun client version that doesn't support passing the spec as CM (i.e. reading it from file), with this change there is no way to make it work except upgrading the client version.
Before making this change constant, most serving graphs would pass as-is in env variable, and the issue may happen only for large serving graphs (which arguable wouldn't work anyway due to their size). Now we're introducing a breaking change that will happen for every serving function deployed using an older client.
We need to give the user an option to keep using env variables for the serving spec using a configuration on the runtime or something similar. Trying to automatically deduce whether the client is old or not is problematic since we don't always know what version of mlrun is included in the image used.

File: server/api/crud/runtimes/nuclio/function.py [hunk 4]
```diff
@@ -212,9 +213,17 @@ def _compile_function_config(
     serving_spec_volume = None
     serving_spec = function._get_serving_spec()
     if serving_spec is not None:
+        # To keep backward comatability, allow passing service spec
+        # via Config Map only for client version higher then 1.7.0
```
Comment (gtopper) [inline server/api/crud/runtimes/nuclio/function.py:218]:
```suggestion
        # via Config Map only for client version 1.7.0 or higher
```

File: server/api/crud/runtimes/nuclio/function.py [hunk 5]
```diff
@@ -213,9 +214,17 @@ def _compile_function_config(
     serving_spec_volume = None
     serving_spec = function._get_serving_spec()
     if serving_spec is not None:
+        # To keep backward comatability, allow passing service spec
+        # via Config Map only for client version 1.7.0 or higher
```
Comment (gtopper) [inline server/api/crud/runtimes/nuclio/function.py]: I guess this needs a TODO about removal in 1.9.0?

File: tests/api/runtimes/test_serving.py [hunk 6]
```diff
@@ -248,6 +248,8 @@ def _setup_serving_spec_in_config_map(self):
     def test_remote_deploy_with_secrets(self, use_config_map):
         if use_config_map:
             self._setup_serving_spec_in_config_map()
+        else:
+            mlrun.mlconf.httpdb.nuclio.serving_spec_env_cutoff = 4096
```
Comment (gtopper) [inline tests/api/runtimes/test_serving.py:252]: Isn't this sticky, potentially affecting tests that run after it?

File: tests/api/api/test_functions.py [hunk 7]
```diff
@@ -481,6 +481,13 @@ def test_tracking_on_serving(
     Validate that `.set_tracking()` configurations are applied to
     a serving function for model monitoring.
     """
+    server.api.utils.singletons.k8s.get_k8s_helper().v1api = unittest.mock.Mock()
+
+    config_map = unittest.mock.Mock()
+    config_map.items = []
+    server.api.utils.singletons.k8s.get_k8s_helper().v1api.list_namespaced_config_map = unittest.mock.Mock(
+        return_value=config_map
+    )
```
Comment (theSaarco) [inline tests/api/api/test_functions.py]: You need to revert these changes at the end of the test. Since you have `monkeypatch` available, best to use it to perform these patching, as it will take care of reverting automatically.

**Review by theSaarco — APPROVED** [review]
LGTM - one small fix needed in the test.

File: tests/api/runtimes/test_serving.py (hunk 6 above)
Comment (alxtkr77) [reply tests/api/runtimes/test_serving.py:252]: It was set to 0 in _setup_serving_spec_in_config_map() before, so I guess not

File: server/api/crud/runtimes/nuclio/function.py (hunk 5 above)
Comment (alxtkr77) [reply server/api/crud/runtimes/nuclio/function.py]: Are we committed for BC only 2 versions back?

Comment (gtopper) [reply server/api/crud/runtimes/nuclio/function.py]: Yup

**Review by gtopper — APPROVED** [review]
